haystack/nodes/sampler/top_p_sampler.py

```diff
@@ -0,0 +1,161 @@
+        :param use_gpu: Whether to use GPU (if available)
+        :param devices: List of torch devices (e.g. cuda:0, cpu, mps) to limit inference to specific devices.
+        :param use_auth_token: The token to use as HTTP bearer authorization for remote files.
+
+        """
+        super().__init__()
+
+        self.top_p = top_p
+        self.top_score_name = top_score_name
+        self.strict = strict
+        self.devices, _ = initialize_device_settings(devices=devices, use_cuda=use_gpu, multi_gpu=True)
+        self.cross_encoder = CrossEncoder(model_name_or_path, device=str(self.devices[0]))
+
+    def predict(self, query: str, documents: List[Document], top_p: Optional[float] = None) -> List[Document]:
+        """
+        Returns a top p filtered list of documents based on the similarity scores between the query and the documents
+        whose cumulative probability is less than or equal to top_p.
+
+        :param query: Query string
+        :param documents: List of Document
+        :param top_p: Cumulative probability threshold for filtering the documents, if not provided, the value of top_p
+        set during TopPSampler initialization will be used
+        :return: List of Document sorted by (desc.) similarity with the query
+        """
+        if top_p is None:
+            top_p = self.top_p
+
+        if not documents:
+            return []
+
+        # prepare the data for the cross encoder
+        query_doc_pairs = [[query, doc.content] for doc in documents]
+
+        # compute the similarity scores for these combinations
+        similarity_scores = self.cross_encoder.predict(query_doc_pairs)
+
+        # Apply softmax normalization to the similarity scores
+        probs = np.exp(similarity_scores) / np.sum(np.exp(similarity_scores))
+
+        # Sort the probabilities and calculate their cumulative sum
+        sorted_probs = np.sort(probs)[::-1]
+        cumulative_probs = np.cumsum(sorted_probs)
+
+        # Find the indices with cumulative probabilities that exceed top_p
+        top_p_indices = np.where(cumulative_probs <= top_p)[0]
+
+        # Map the selected indices back to their original indices
+        original_indices = np.argsort(probs)[::-1][top_p_indices]
+        # and select the top_p responses
+        selected_docs = [documents[i] for i in original_indices]
+
+        # if strict is set to False, low p resulted in no documents being selected, then
+        # return at least one document
+        if not selected_docs and not self.strict:
+            highest_prob_indices = np.argsort(probs)[::-1]
+            selected_docs = [documents[highest_prob_indices[0]]]
+
+        # include prob scores in the results
+        if self.top_score_name:
+            for idx, doc in enumerate(selected_docs):
+                doc.meta[self.top_score_name] = "{:.2f}".format(sorted_probs[idx])
```

Comment:
I'd say the formatting should be done by the application. Returning the scores with high precision here is fine.